gadgets/src/curves/templates/bls12/g2.rs

```diff
@@ -282,3 +323,74 @@ impl<P: Bls12Parameters> ConditionalEqGadget<<P as Bls12Parameters>::Fp> for G2P
         unimplemented!()
     }
 }
+
+impl<P: Bls12Parameters> SumGadget<<P as Bls12Parameters>::Fp> for G2PreparedGadget<P> {
+    fn zero<CS: ConstraintSystem<<P as Bls12Parameters>::Fp>>(mut cs: CS) -> Result<Self, SynthesisError> {
+        let zero = Fp2G::<P>::zero(cs.ns(|| "zero coeff entry element"))?;
+
+        let bit_iterator = BitIteratorBE::new(P::X);
+        let mut ell_coeffs = Vec::with_capacity(bit_iterator.len());
+
+        for i in bit_iterator.skip(1) {
+            ell_coeffs.push((zero.clone(), zero.clone()));
+            if i {
+                ell_coeffs.push((zero.clone(), zero.clone()));
+            }
+        }
+
+        Ok(Self { ell_coeffs })
+    }
+
+    fn sum<CS: ConstraintSystem<<P as Bls12Parameters>::Fp>>(
+        mut cs: CS,
+        elems: &[Self],
+    ) -> Result<Self, SynthesisError> {
+        assert!(!elems.is_empty());
```

Comment:
is this assertion safe to make?

Reply:
Changed to another implementation that allows an empty `elems`:

```
        let mut res = Self::zero(cs.ns(|| "zero"))?;
        for (i, elem) in elems.iter().enumerate() {
            for (j, (l_coeff, r_coeff)) in elem.ell_coeffs.iter().enumerate() {
                res.ell_coeffs[j].0 = res[j].0.add(cs.ns(|| format!("sum_{}_{}_entry_0", i, j)), l_coeff)?;
                res.ell_coeffs[j].1 = res[j].1.add(cs.ns(|| format!("sum_{}_{}_entry_1", i, j)), r_coeff)?;
            }
        }

        Ok(res)
```